### ISSUE TITLE
[LSP] diagnostic severity_sort no longer works

### ISSUE THREAD

**Issue by ray-x** [neovim/neovim#15685]
### Neovim version (nvim -v)

v0.6.0-dev+284-gc2a65921d

### Vim (not Nvim) behaves the same?

no

### Operating system/version

macOS

### Terminal name/version

kitty

### $TERM environment variable

xterm-kitty

### Installation

source code build

### How to reproduce the issue

in navigator.lua diagnostic handler, setup severity_sort.
Setup  vim.lsp.handlers["textDocument/publishDiagnostics"]  with following code:

```lua
local diagnostic_cfg = {
  underline = true,
  virtual_text = {spacing = 3, prefix = 'E'},
  signs = true,
  update_in_insert = false,
  severity_sort = function(a, b)
    return a.severity < b.severity
  end
}

 vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
    vim.lsp.diagnostic.on_publish_diagnostics,  diagnostic_cfg
  )

```
I got an error on `Error executing vim.schedule lua callback: ...k/packer/opt/navigator.lua/lua/navigator/diagnostics.lua:242: attempt to index local 'a' (a
 number value)   `

The traceback:

```
       /usr/local/share/nvim/runtime/lua/vim/diagnostic.lua:53: in function 'resolve_optional_value'
	/usr/local/share/nvim/runtime/lua/vim/diagnostic.lua:98: in function 'get_resolved_options'
	/usr/local/share/nvim/runtime/lua/vim/diagnostic.lua:974: in function 'show'
	/usr/local/share/nvim/runtime/lua/vim/diagnostic.lua:469: in function 'config'
	/usr/local/share/nvim/runtime/lua/vim/lsp/diagnostic.lua:198: in function 'on_publish_diagnostics'
	...k/packer/opt/navigator.lua/lua/navigator/diagnostics.lua:167: in function 'handler'
	/usr/local/share/nvim/runtime/lua/vim/lsp.lua:681: in function 'cb'

```

### Expected behavior

there should be no error

### Actual behavior

nil de-reference

```
attempt to index local 'a' (a
 number value) 
```

**Comment by clason** [comment]
The API change in https://github.com/neovim/neovim/pull/15585 is breaking; all functions passed via `opts` now need to accept a  `namespace` and `bufnr`:
```
 ---@note Each of the configuration options below accepts one of the following:
  ---         - `false`: Disable this feature
  ---         - `true`: Enable this feature, use default settings.
  ---         - `table`: Enable this feature with overrides.
  ---         - `function`: Function with signature (namespace, bufnr) that returns any of the above.
```

**Comment by gpanders** [comment]
I am actually surprised this worked before. It seems like it may have been undocumented behavior. From the old LSP docs:

```
                                 *vim.lsp.diagnostic.on_publish_diagnostics()*
on_publish_diagnostics({_}, {result}, {ctx}, {config})
                |lsp-handler| for the method "textDocument/publishDiagnostics"

                Note:
                    Each of the configuration options accepts:
                    • `false` : Disable this feature
                    • `true` : Enable this feature, use default settings.
                    • `table` : Enable this feature, use overrides.
                    • `function`: Function with signature (bufnr, client_id) that
                      returns any of the above.>
```

I don't see anything in the old docs that indicates that `severity_sort` can take an actual sorting function.

In any case, I'll look into fixing this as well as documenting it. Thanks for the bug report.

**Comment by ray-x** [comment]
Thanks!

**Comment by gpanders** [comment]
I don't think we actually need the ability to specify a function. We just need to be able to provide an "ascending" or "descending" sort order.

**Comment by gpanders** [comment]
Should be fixed in #15696

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/15696
